src/serac/physics/materials/thermal_functional_material.hpp

```diff
@@ -0,0 +1,278 @@
+  double specific_heat_capacity_ = 1.0;
+
+  /// Constant thermal conductivity
+  tensor<double, dim, dim> conductivity_ = {{{1.0, 0.0, 0.0}, {0.0, 1.0, 0.0}, {0.0, 0.0, 1.0}}};
+};
+
+// Use SFINAE to add static assertions checking if the given thermal material type is acceptable
+template <typename T, int dim, typename = void>
+struct has_density : std::false_type {
+};
+
+template <typename T, int dim>
+struct has_density<T, dim, std::void_t<decltype(std::declval<T&>().density(std::declval<tensor<double, dim>&>()))>>
+    : std::true_type {
+};
+
+template <typename T, int dim, typename = void>
+struct has_specific_heat_capacity : std::false_type {
+};
+
+template <typename T, int dim>
+struct has_specific_heat_capacity<T, dim,
+                                  std::void_t<decltype(std::declval<T&>().specificHeatCapacity(
+                                      std::declval<tensor<double, dim>&>(), std::declval<tensor<double, 1>&>()))>>
+    : std::true_type {
+};
+
+template <typename T, int dim, typename = void>
+struct has_thermal_flux : std::false_type {
+};
+
+template <typename T, int dim>
+struct has_thermal_flux<
+    T, dim,
+    std::void_t<decltype(std::declval<T&>()(std::declval<tensor<double, 1>&>(), std::declval<tensor<double, dim>&>()))>>
+    : std::true_type {
+};
+
+/// Constant thermal source model
+struct ConstantSource {
+  /// The constant source
+  double source_;
+
+  /**
+   * @brief Evaluation function for the constant thermal source model
+   *
+   * @tparam T1 type of the temperature
+   * @tparam T2 type of the temperature gradient
+   * @tparam dim The dimension of the problem
+   * @return The thermal source value
+   */
+  template <typename T1, typename T2, int dim>
+  SERAC_HOST_DEVICE double operator()(tensor<double, dim>& /* x */, double /* t */, T1& /* u */, T2& /* du_dx */) const
+  {
+    return source_;
+  }
+};
+
+// Use SFINAE to add static assertions checking if the given thermal source type is acceptable
+template <typename T, int dim, typename = void>
+struct has_thermal_source : std::false_type {
+};
+
+template <typename T, int dim>
+struct has_thermal_source<
+    T, dim,
+    std::void_t<decltype(std::declval<T&>()(std::declval<tensor<double, dim>&>(), std::declval<double>(),
+                                            std::declval<tensor<double, 1>&>(), std::declval<tensor<double, dim>&>()))>>
+    : std::true_type {
+};
+
+/// Constant thermal flux boundary model
+struct FluxBoundary {
+  /// The constant flux applied to the boundary
+  double flux_;
+
+  /**
+   * @brief Evaluation function for the thermal flux on a boundary
+   *
+   * @tparam T1 Type of the normal vector
+   * @tparam T2 Type of the temperature
+   * @tparam dim The dimension of the problem
+   * @return The flux applied to the boundary
+   */
+  template <typename T1, typename T2, int dim>
+  SERAC_HOST_DEVICE double operator()(tensor<double, dim>& /* x */, T1& /* n */, T2& /* u */) const
+  {
+    return flux_;
+  }
+};
```

Comment:
this is beyond the scope of this PR, but it might be worth adding a feature to Functional that allows it to do surface / domain integrals that do not depend on the trial spaces, so we won't be interpolating the finite element fields and then just discarding the values in the q-function.